packages/search-in-workspace/src/node/ripgrep-search-in-workspace-server.ts

```diff
@@ -290,6 +339,56 @@ export class RipgrepSearchInWorkspaceServer implements SearchInWorkspaceServer {
         return Promise.resolve(searchId);
     }
 
+    protected resolveSearchPaths(rootPaths: string[], opts: SearchInWorkspaceOptions | undefined): string[] {
+        if (!opts || !opts.include) {
+            return rootPaths;
+        }
+
+        const searchPaths: string[] = [];
+        const patternPaths: string[] = [];
+        opts.include.forEach(pattern => {
+            rootPaths.forEach(root => {
+                let searchPath = path.join(root, pattern);
+                let isPatternAValidPath = false;
+                if (fs.existsSync(searchPath)) {
+                    isPatternAValidPath = true;
+                } else if (fs.existsSync(pattern)) {
+                    searchPath = pattern;
+                    isPatternAValidPath = true;
+                } else if (pattern.endsWith('**')) {
+                    const patternPath = this.includeSubDirectoriesToPath(pattern);
+                    if (fs.existsSync(patternPath)) {
+                        searchPath = patternPath;
+                        isPatternAValidPath = true;
+                    }
+                }
+
+                if (isPatternAValidPath) {
+                    searchPaths.push(searchPath);
+                    patternPaths.push(pattern);
+                }
+            });
+        });
+
+        // Exclude include file patters that were successfully translated to search paths
+        opts.include = opts.include.filter(item => !patternPaths.includes(item));
+        return searchPaths.length > 0 ? searchPaths : rootPaths;
+    }
+
+    protected includeSubDirectoriesToPath(pattern: string): string {
```

Comment:
Can you document what this method is attempting to achieve?
Can it not be generalized rather than only for `include`?

It might also help to unit-test it to verify it is behaving as intended under different use-cases.

Reply:
Done, the newest patch set reduced the size of the methods added, I have added some doc info and additional test cases were created.